apps/collections/api/views.py

```diff
@@ -12,3 +24,63 @@ class CollectionViewSet(ModelViewSet):
     lookup_field = 'slug'
     serializer_class = CollectionSerializer
     permission_classes = [permissions.IsAuthenticatedOrReadOnly]
+
+
+class CollectionEditItemsView(APIView):
+    """Edits My Collection items."""
+
+    my_collection_name = _('My Collection')
+
+    permission_classes = [permissions.IsAuthenticated]
+
+    def post(self, request: 'Request', *args, **kwargs):
+        """Apply items to the collection."""
+        data = request.data
+        sources = data.get('sources', [])
+        propositions = data.get('propositions', [])
+        entities = data.get('entities', [])
+        quotes = data.get('quotes', [])
+
+        collection_args = {'title': self.my_collection_name, 'creator': request.user}
+        try:
+            collection = Collection.objects.get(**collection_args)
+        except Collection.DoesNotExist:
+            collection = Collection.objects.create(**collection_args)
+
+        self.apply_items(
+            collection,
+            {
+                'sources': sources,
+                'propositions': propositions,
+                'entities': entities,
+                'quotes': quotes,
+            },
+        )
+
+        return Response(CollectionSerializer(collection).data)
+
+    def apply_items(self, collection: Collection, items: dict):
+        raise NotImplementedError
+
+
+class CollectionAddItemsView(CollectionEditItemsView):
+    """Add items to the collection."""
+
+    def apply_items(self, collection: Collection, items: dict):
+        for module_type in COLLECTION_MODULE_TYPES:
+            try:
+                objects = items.get(module_type, [])
+                if objects:
+                    getattr(collection, module_type).add(*objects)
+            except IntegrityError:
+                raise ValidationError(f'Some {module_type} were not found in the database')
```

Comment:
This could be moved to parent class in the future so CollectionRemoveItemsView can use similar validation logic too.